File: backend/services/Order/OrderService.js
const OrderRepository = require('../../repository/OrderRepository');
const {NotFoundError, ForbiddenError, ValidationError, OrderError} = require('../../utils/Errors');
const OrderFactory = require('./OrderFactory');

class OrderService {
    static async createOrder(orderData) {
        try {
            const order = await OrderFactory.createOrder(orderData);
            const validationErrors = order.validate_order();
            
            if (validationErrors.length > 0) {
                throw new ValidationError(`訂單驗證失敗: ${JSON.stringify(validationErrors)}`);
            }

            return order;
        } catch (error) {
            throw new OrderError(`創建訂單失敗: ${error.message}`);
        }
    }
    static async cancelOrder(order_id, user_id) {
        try {
            // 1. 檢查訂單是否存在且屬於該用戶
            const orderResult = await OrderRepository.get({ order_id });
            if (!orderResult.rows.length) {
                throw new NotFoundError('訂單不存在');
            }

            const order = orderResult.rows[0];
            if (order.user_id !== user_id) {
                throw new ForbiddenError('無權限取消此訂單');
            }

            // 2. 檢查訂單是否可以取消
            if (order.status !== 'pending' && order.status !== 'partial') {
                throw new ValidationError('訂單狀態不允許取消');
            }

            // 3. 從 OrderBook 中移除訂單
            const orderbook = OrderBookService.getOrderBook();
            await orderbook.removeOrder(order_id);
            
            // 4. 更新訂單狀態
            await OrderRepository.update(order_id, { 
                status: 'cancelled' 
            });

            return true;

        } catch (error) {
            throw new OrderError(`取消訂單失敗: ${error.message}`);
        }
    }

    static async getOrderStatus(order_id, user_id) {
        try {
            const orderResult = await OrderRepository.get({ order_id });
            if (!orderResult.rows.length) {
                throw new NotFoundError('訂單不存在');
            }

            const order = orderResult.rows[0];
            if (order.user_id !== user_id) {
                throw new ForbiddenError('無權限查看此訂單');
            }

            return {
                order_id,
                status: order.status,
                filled_quantity: order.quantity - order.remaining_quantity,
                remaining_quantity: order.remaining_quantity,
                price: order.price,
                created_at: order.created_at
            };

        } catch (error) {
            throw new OrderError(`獲取訂單狀態失敗: ${error.message}`);
        }
    }

}

module.exports = OrderService;

File: backend/services/order/orderService.py
from typing import Dict, Any
from backend.repository.orderRepository import OrderRepository
from utils.errors import NotFoundError, ForbiddenError, ValidationError, OrderError
from backend.models.order.orderFactory import OrderFactory
from backend.services.orderbook.orderbookService import OrderBookService
from backend.models.order.order import OrderBase
from backend.database.database import Database
class OrderService:
    @staticmethod
    async def create_order(order_data: Dict[str, Any]) -> Any:
        """
        創建訂單
        
        Args:
            order_data: 訂單數據
            
        Returns:
            創建的訂單對象
            
        Raises:
            OrderError: 創建訂單失敗時
            ValidationError: 訂單驗證失敗時
        """
        try:
            async with Database.transaction() as client:
                order = await OrderFactory.create_order(order_data)
                validation_errors = order.validate_order()
                
                if validation_errors:
                    raise ValidationError(f"訂單驗證失敗: {validation_errors}")
                order_id = await OrderRepository.insert(order, client)
                order.order_id = order_id
            return order
            
        except Exception as error:
            raise error

    @staticmethod
    async def cancel_order(order_id: int, user_id: int) -> bool:
        """
        取消訂單
        
        Args:
            order_id: 訂單ID
            user_id: 用戶ID
            
        Returns:
            bool: 是否成功取消
            
        Raises:
            NotFoundError: 訂單不存在時
            ForbiddenError: 無權限時
            ValidationError: 訂單狀態不允許取消時
            OrderError: 取消訂單失敗時
        """
        try:
            orders = await OrderRepository.get({"order_id": order_id})

            if len(orders)==0:
                raise NotFoundError("訂單不存在")
            order = orders[0]
            if order["user_id"] != user_id:
                raise ForbiddenError("無權限取消此訂單")
            
            # 3. 從 OrderBook 中移除訂單
            await OrderBookService.remove_order(order_id)

            order = dict(order)

            # 4. 更新訂單狀態
            await OrderRepository.update(
                order['order_id'],
                {"status": "canceled"}
            )

            return True

        except Exception as error:
            raise OrderError(f"取消訂單失敗: {str(error)}")

    @staticmethod
    async def get_order_status(order_id: int, user_id: int) -> Dict[str, Any]:
        """
        獲取訂單狀態
        
        Args:
            order_id: 訂單ID
            user_id: 用戶ID
            
        Returns:
            Dict[str, Any]: 訂單狀態信息
            
        Raises:
            NotFoundError: 訂單不存在時
            ForbiddenError: 無權限時
            OrderError: 獲取狀態失敗時
        """
        try:
            order_result = await OrderRepository.get({"order_id": order_id})
            if not order_result.rows:
                raise NotFoundError("訂單不存在")

            order = order_result.rows[0]
            if order["user_id"] != user_id:
                raise ForbiddenError("無權限查看此訂單")

            return {
                "order_id": order_id,
                "status": order["status"],
                "filled_quantity": order["quantity"] - order["remaining_quantity"],
                "remaining_quantity": order["remaining_quantity"],
                "price": order["price"],
                "created_at": order["created_at"]
            }

        except Exception as error:
            raise OrderError(f"獲取訂單狀態失敗: {str(error)}")